### PR TITLE
USD Keyframe Animations

### PR DESCRIPTION
**Pull request by AMZN-Gene**
- Imports animation data from TinyUSDZ to support keyframe animations
- Templatized tinyUsdzMat4ToAiMat4 function to accept TinyUSDZ float and double matrix; Templatized also satisfies GitHub's quality gate check for duplicate code

https://github.com/user-attachments/assets/13771c23-f2d7-454b-ad21-c79a2a6f3c1b

**Tested By**
Loading toy_drummer_idle into AssimpViewer and seeing animation play
 - Note: I had to re-export Apple's original ToyDrummer inside Blender, otherwise the [TinyUSDZ bug with bone weights](https://github.com/lighttransport/tinyusdz/issues/206) appears. File here: 
[toy_drummer_idle_reexport.zip](https://github.com/user-attachments/files/17604624/toy_drummer_idle_reexport.zip)

**Future Work**
- Fix TinyUSDZ to correctly load bone weights, https://github.com/lighttransport/tinyusdz/issues/206
- Support multiple animations https://github.com/lighttransport/tinyusdz/issues/122
- For applying multiple animations to an object, the other option is to store animations separately inside separate USDs; one per animation, then copying the animation-aiScenes into the original aiScene. While possible, this method requires a dummy mesh to be included in the animation USD. It would be good if the dummy mesh wasn't required.

### CODE REVIEW

**Review by moudgils — APPROVED** [review]
Looks good

**Comment by AMZN-Gene** [comment]
@tellypresence would love your review if you get a chance, thanks!

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp [hunk 1]
```diff
@@ -223,6 +224,124 @@ void USDImporterImplTinyusdz::InternReadFile(
...
+            newAiNodeAnim->mNumPositionKeys = positionKeys.size();
+            newAiNodeAnim->mPositionKeys = new aiVectorKey[newAiNodeAnim->mNumPositionKeys];
+            std::move(positionKeys.begin(), positionKeys.end(), newAiNodeAnim->mPositionKeys);
+
+            newAiNodeAnim->mNumRotationKeys = rotationKeys.size();
+            newAiNodeAnim->mRotationKeys = new aiQuatKey[newAiNodeAnim->mNumRotationKeys];
+            std::move(rotationKeys.begin(), rotationKeys.end(), newAiNodeAnim->mRotationKeys);
+
+            newAiNodeAnim->mNumScalingKeys = scalingKeys.size();
+            newAiNodeAnim->mNumScalingKeys = scalingKeys.size();
```
Comment (spham-amzn) [inline code/AssetLib/USD/USDLoaderImplTinyusdz.cpp:337]: This looks like a duplicated line

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp [hunk 2]
```diff
@@ -223,6 +224,124 @@ void USDImporterImplTinyusdz::InternReadFile(
...
+        const auto &animation = render_scene.animations[animationIndex];
+
+        auto newAiAnimation = new aiAnimation();
+        pScene->mAnimations[animationIndex] = newAiAnimation;
+
+        newAiAnimation->mName = animation.abs_path;
+
+        // each channel affects a node (joint)
+        newAiAnimation->mNumChannels = animation.channels_map.size();
+        newAiAnimation->mChannels = new aiNodeAnim *[newAiAnimation->mNumChannels];
```
Comment (spham-amzn) [inline code/AssetLib/USD/USDLoaderImplTinyusdz.cpp:248]: Do you need to check for a non-empty collection of channels here as well?

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp [hunk 3]
```diff
@@ -223,6 +224,124 @@ void USDImporterImplTinyusdz::InternReadFile(
...
+                case tinyusdz::tydra::AnimationChannel::ChannelType::Rotation:
+                    if (animChannel.rotations.static_value.has_value()) {
+                        rotationKeys.push_back(aiQuatKey(0, tinyUsdzQuatToAiQuat(animChannel.rotations.static_value.value())));
+                    }
+                    for (const auto &rotationAnimSampler : animChannel.rotations.samples) {
+                        if (rotationAnimSampler.t > newAiAnimation->mDuration) {
+                            newAiAnimation->mDuration = rotationAnimSampler.t;
+                        }
+
+                        rotationKeys.push_back(aiQuatKey(rotationAnimSampler.t, tinyUsdzQuatToAiQuat(rotationAnimSampler.value)));
```
Comment (spham-amzn) [inline code/AssetLib/USD/USDLoaderImplTinyusdz.cpp:271]: suggestion: maybe use `emplace_back` instead?
```suggestion
                        rotationKeys.emplace_back(rotationAnimSampler.t, tinyUsdzQuatToAiQuat(rotationAnimSampler.value));
```

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp [hunk 4]
```diff
@@ -278,6 +401,8 @@ void USDImporterImplTinyusdz::verticesForMesh(
         std::vector<const tinyusdz::tydra::SkelNode *> skeletonNodes;
         skeletonNodes.push_back(&render_scene.skeletons[skeleton_id].root_node);
         for (int i = 0; i < skeletonNodes.size(); ++i) {
+            const tinyusdz::tydra::SkelNode *skeletonNode = skeletonNodes[i];
```
Comment (tellypresence) [inline code/AssetLib/USD/USDLoaderImplTinyusdz.cpp:404]: Please remove this line; `skeletonNode` is unused and will break the build in environments which treat warnings as errors

**Review by tellypresence — CHANGES_REQUESTED** [review]

**Comment by tellypresence** [comment]
Confirm there is animation, but appears to be incorrect hierachy; drum and winding key seem to be parented to left arm and they oscillate as a unit
apple_toy_drummer_idle.usdz

**Comment by AMZN-Gene** [comment]
> For Apple drummer model, confirm there is animation, but appears to be incorrect hierachy; drum and winding key seem to be parented to left arm and they oscillate as a unit apple_toy_drummer_idle.usdz [ apple_toy_drummer_idle.usdz ](https://private-user-images.githubusercontent.com/26958898/386930617-04bd96cf-9854-4c7a-a7d8-6683ae8b3b5e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5MzA2MTctMDRiZDk2Y2YtOTg1NC00YzdhLWE3ZDgtNjY4M2FlOGIzYjVlLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzOWIwNjk4NmI5OTJiMGY2NTI2NGQ2MmRlZWQ3MzQ1YjliMmUwNjA3NjBjODU0YmZlYmQ5Y2Y3ZTk0ZDhiYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K7Q9XLjt9HtDWk_obl6hKM96gdfXVZT9FHBWcYtccB8) [ ](https://private-user-images.githubusercontent.com/26958898/386930617-04bd96cf-9854-4c7a-a7d8-6683ae8b3b5e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5MzA2MTctMDRiZDk2Y2YtOTg1NC00YzdhLWE3ZDgtNjY4M2FlOGIzYjVlLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzOWIwNjk4NmI5OTJiMGY2NTI2NGQ2MmRlZWQ3MzQ1YjliMmUwNjA3NjBjODU0YmZlYmQ5Y2Y3ZTk0ZDhiYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K7Q9XLjt9HtDWk_obl6hKM96gdfXVZT9FHBWcYtccB8)
> 
> Apple's robot also has animation but doesn't appear to be correct apple_robot_walk_idle.usdz [ apple_robot_walk_idle.usdz ](https://private-user-images.githubusercontent.com/26958898/386931374-dc9e6ec5-ac55-4331-9473-5e28e1c8320c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5MzEzNzQtZGM5ZTZlYzUtYWM1NS00MzMxLTk0NzMtNWUyOGUxYzgzMjBjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2MDNiN2U4NzUwYzFiNjIxNjYzMWYxNTlhZjdjZWE2NWViYTI5NjYwOGRjYTViZDU4NDJjOWI1NjAxMjI1MjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lkFXNhrlBfP3xJVWoHI_HxyYN6b-SdcjwLmlkm5HPjg) [ ](https://private-user-images.githubusercontent.com/26958898/386931374-dc9e6ec5-ac55-4331-9473-5e28e1c8320c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5MzEzNzQtZGM5ZTZlYzUtYWM1NS00MzMxLTk0NzMtNWUyOGUxYzgzMjBjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2MDNiN2U4NzUwYzFiNjIxNjYzMWYxNTlhZjdjZWE2NWViYTI5NjYwOGRjYTViZDU4NDJjOWI1NjAxMjI1MjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lkFXNhrlBfP3xJVWoHI_HxyYN6b-SdcjwLmlkm5HPjg)
> 
> Animation for these models looks better: Re-exported Apple drummer (.zip file attached above) apple_toy_drummer_idle_usdz_reexport [ apple_toy_drummer_idle_usdz_reexport ](https://private-user-images.githubusercontent.com/26958898/386987658-02b742ad-41ad-4d8c-b783-c402ad83103b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5ODc2NTgtMDJiNzQyYWQtNDFhZC00ZDhjLWI3ODMtYzQwMmFkODMxMDNiLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMTBmYmZkYWJlMGFmN2E1Nzc2NGI5YmRjNmI0NDkwYTBhYTljM2YxNzg0MGNlNTIyNDg0YWMzOTA0NDYxZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nZAtMK_RDJvQNTxwLGtL8Ljzm6Tdm7MQV2ahBLVv-tg) [ ](https://private-user-images.githubusercontent.com/26958898/386987658-02b742ad-41ad-4d8c-b783-c402ad83103b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5ODc2NTgtMDJiNzQyYWQtNDFhZC00ZDhjLWI3ODMtYzQwMmFkODMxMDNiLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMTBmYmZkYWJlMGFmN2E1Nzc2NGI5YmRjNmI0NDkwYTBhYTljM2YxNzg0MGNlNTIyNDg0YWMzOTA0NDYxZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nZAtMK_RDJvQNTxwLGtL8Ljzm6Tdm7MQV2ahBLVv-tg)
> 
> Brainstem (1121 bones in 59 meshes; probably actually 19 bones (1121 / 59 = 19)) brainstem.usdz [ brainstem.usdz ](https://private-user-images.githubusercontent.com/26958898/386986062-32fa2cf0-4068-4bf6-8289-fddac0d6ac6e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5ODYwNjItMzJmYTJjZjAtNDA2OC00YmY2LTgyODktZmRkYWMwZDZhYzZlLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjZjk5MWQ5ZTk3MWIzNThlMDA5ODQ5NDdiZWY5YzJiYjFhZDFhN2FlOTVkZDJhODA1NzM2NGM3MjE0MmQyOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JNNauHQMNEEkA6cDJIWG_D0zmuHavkMLwekhCNK-VIY) [ ](https://private-user-images.githubusercontent.com/26958898/386986062-32fa2cf0-4068-4bf6-8289-fddac0d6ac6e.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5ODYwNjItMzJmYTJjZjAtNDA2OC00YmY2LTgyODktZmRkYWMwZDZhYzZlLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjZjk5MWQ5ZTk3MWIzNThlMDA5ODQ5NDdiZWY5YzJiYjFhZDFhN2FlOTVkZDJhODA1NzM2NGM3MjE0MmQyOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JNNauHQMNEEkA6cDJIWG_D0zmuHavkMLwekhCNK-VIY)
> 
> CesiumMan (19 bones in 1 mesh) CesiumMan.usdz [ CesiumMan.usdz ](https://private-user-images.githubusercontent.com/26958898/386986686-437434f4-d69c-4f72-b7ec-046b350df338.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5ODY2ODYtNDM3NDM0ZjQtZDY5Yy00ZjcyLWI3ZWMtMDQ2YjM1MGRmMzM4LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiNTBjMzk1MDhmOWFmNmY0YmIxMThjYThmNDY1MTEyN2IwNzBhNmY3M2Y2MTU5YzYwNWUzYzFhNmUzMzlmMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.foxrku-KKG5w9jAgNwkAw4flbGd5-9cW_Y8cYPS085A) [ ](https://private-user-images.githubusercontent.com/26958898/386986686-437434f4-d69c-4f72-b7ec-046b350df338.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTA3MzcsIm5iZiI6MTczMjAxMDQzNywicGF0aCI6Ii8yNjk1ODg5OC8zODY5ODY2ODYtNDM3NDM0ZjQtZDY5Yy00ZjcyLWI3ZWMtMDQ2YjM1MGRmMzM4LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEwMDAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiNTBjMzk1MDhmOWFmNmY0YmIxMThjYThmNDY1MTEyN2IwNzBhNmY3M2Y2MTU5YzYwNWUzYzFhNmUzMzlmMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.foxrku-KKG5w9jAgNwkAw4flbGd5-9cW_Y8cYPS085A)

i ran into this as well. I had to re-export Apple's original ToyDrummer inside Blender, otherwise the https://github.com/lighttransport/tinyusdz/issues/206 appears.

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp (hunk 4 above)
Comment (AMZN-Gene) [reply code/AssetLib/USD/USDLoaderImplTinyusdz.cpp:404]: Thank you! Fixed!

**Comment by AMZN-Gene** [comment]
So would the TinyUSDZ animation bug block this pull request? Assimp users may get confused when basic animations like ToyDrummer don't work as expected. Likewise, we could check this in, and make a separate PR to grab latest Tiny once their fix is in and everything will just work.

https://github.com/lighttransport/tinyusdz/issues/206

**Review by kimkulling — APPROVED** [review]
Looks fine, let's try it out.

**Comment by kimkulling** [comment]
Merged, thanks a lot for your contribution.

**Pull request merged**